build/templates/_converters.py.mako

```diff
@@ -158,6 +158,98 @@ def convert_repeated_capabilities_without_prefix(repeated_capability):
+            --> ['Dev1/0', 'Dev1/1', 'Dev1/2']
+        - convert_channels_repeated_capabilities('Dev1/0:2', ['Dev1/0', 'Dev1/1', 'Dev1/2', 'Dev1/3'])
+            --> ['Dev1/0', 'Dev1/1', 'Dev1/2']
+        - convert_channels_repeated_capabilities('Dev1/0:2,4', ['Dev1/0', 'Dev1/1', 'Dev1/2', 'Dev1/3', 'Dev1/4', 'Dev1/5'])
+            --> ['Dev1/0', 'Dev1/1', 'Dev1/2', 'Dev1/4']
+        - convert_channels_repeated_capabilities('4,Dev1/1:2', ['Dev1/1', 'Dev1/2', 'Dev1/4'])
+            --> ['Dev1/4', 'Dev1/1', 'Dev1/2']
+        - convert_channels_repeated_capabilities('Dev1/4,Dev1/2,Dev1/3', ['Dev1/2', 'Dev1/3', 'Dev1/4'])
+            --> ['Dev1/4', 'Dev1/2', 'Dev1/3']
+        - convert_channels_repeated_capabilities('Dev1/1,Dev2/2', ['Dev1/0', 'Dev1/1', 'Dev1/2', 'Dev1/3', 'Dev2/0', 'Dev2/1', 'Dev2/2', 'Dev2/3'])
+            --> ['Dev1/1', 'Dev2/2']
+
+    Args:
+        channels_repeated_capability (str) - refer to _convert_repeated_capabilities() for the
+            supported formats (this string is expected to be used as the index of session.channels)
+
+        session_channel_names (list of str) - names of all the channels in the session, as returned
+            by get_channel_names() (if they have a prefix, which ends with '/', it would be added to
+            any of the expanded items if it does not already have one)
+
+    Returns:
+        channel_names (list of str) - list of strings of the expanded channels repeated capability
+            items (essentially the channel names)
+    '''
+    channels_repeated_capability = channels_repeated_capability.strip()
+    if channels_repeated_capability == '':
+        return session_channel_names
+
+    # Use one of the session_channel_names to deduce if the session was opened with
+    # independent_channels set to True (in that case, the channel name would be fully-qualified) or
+    # False (in that case, the channel name would not have any instrument prefix).
+    session_channel_name = session_channel_names[0]
+    if '/' in session_channel_name:
+        # Split the comma-delimited channels repeated capabilities (if any) into groups with at most
+        # one prefix each and expand their ranges (if any) accordingly
+        repeated_capabilities = []
+        for rep_cap in channels_repeated_capability.split(','):
+            repeated_capabilities.extend(convert_single_group_repeated_capabilities(rep_cap))
+        # If there is any channels repeated capabilities without prefix, the session must have only
+        # one instrument, so just get the prefix from session_channel_name and add to all of the
+        # repeated capabilities that are without prefix
+        default_prefix = session_channel_name[:session_channel_name.find('/') + 1]
+        return [
+            rep_cap if '/' in rep_cap else default_prefix + rep_cap
+            for rep_cap in repeated_capabilities
+        ]
+    return convert_repeated_capabilities(channels_repeated_capability, '')
```

Comment:
I think we can simplify all this logic and get rid of the helper function using `str.rpartition`..

Here's what I'm thinking:
```
# Rule 1: If `all_channels_in_session` is fully-qualified then returned channel names should be
#         fully-qualified. Else returned channel names should not be fully-qualified.
# Rule 2: If any channel in the input is not fully-qualified, but we need to return fully-qualified
#         channels because of Rule 1, then use the channel qualifier obtained from
#         `all_channels_in_session`. This can only happen on a single-instrument session, so all
#         the channel qualifiers are the same and we pick the first one.

instrument, separator, channel = all_channels_in_session[0].rpartition('/')
default_channel_qualifier = instrument + separator

expanded_channel_list = []
for token in channels_string.split(','):
    instrument, separator, channel = token.rpartition('/')
    channel_qualifier = instrument + separator if instrument else default_channel_qualifier
    expanded_channel_list.extend(convert_repeated_capabilities(channel, instrument + separator))

return expanded_channel_list
```

Reply:
Note the use of `strip` and `lower` to make FQCN be in canonical form.

@marcoskirsch, to make the channel names canonical, we can always return all-lower-case or do we need to match what is returned by `get_channel-names`? (which follows the casing passed into resource name during init?)

Reply:
Thanks for the suggestion. I have combined the two converter functions into one single function `expand_channel_string()` in my latest commit. This function takes care of removing the "valid whitespaces" and returning canonical channel names (based on what are returned by `get_channel_names()`; in fact, the channel names returned by this new function would be a subset of the channel names returned by `get_channel_names()`).

Reply:
> or do we need to match what is returned by get_channel-names? 

This is the "correct" behavior in my opinion.